### PR TITLE
Vs find dialog style

### PR DESCRIPTION
**Pull request by ppatpat**
1) Changes the "Find" dialog box behavior like VS
[CTRL]-F, 
[CTRL]-H, 
[CTRL]-[SHIFT]-F 
behavior is changed mimicking VS IDE style.
a) Only one dialog can be displayed at the time.
b) Marked text is automatically loaded as text to find

2) Attaches to the tab menu an option giving the full path of the open file.
Very handy when writing/commenting about code.

### CODE REVIEW

**Comment by elextr** [comment]
You have commits for different things in the one PR, please separate them into different PRs.

**Comment by ppatpat** [comment]
Sorry; I've added both mods together because they are all VS features.
Now I'm not finding an easy way to split the PR from the github WEB interface.
if you can help just let me know.

**Comment by elextr** [comment]
You can leave them together if you wish, but although the first simple enhancement of copy filename to clipboard is likely to be fairly uncontroversial the later parts look like they would be more controversial, so they may delay the original change.

**Comment by elextr** [comment]
I might add that as most of the Geany devs do *not* use windows, you might be better off if you actually described what the changes are, and what their benefits are, rather than "like VS" since that might be meaningless to non-VS users and also their response might be "who cares if its the same as VS, we don't use it".

**Comment by ppatpat** [comment]
Hi there, please believe me; it is not laziness, I just can't find a neat way to separate one PR in 2 PR.
I'm going to describe now in detail both modifications.

Please 1st let me say I'm a guy familiar with VS but I'm using Geany exclusively under Linux, I find these mods very useful when using Geany under Linux but surely they will also be enjoyed by people using Geany under Windows.

Mod 1)
Changes the Find dialog boxes behavior like VS. 
This mod alters the files src/search.c and src/keybindings.c

Before the patch:
 a) The short-cuts
[CTRL]-F, 
[CTRL]-H, 
[CTRL]-[SHIFT]-F 
open the dialogs "Find", "Find & Replace", and "Find in files".
All these dialog boxes can coexist "simultaneously" on top of the current file what produces a highly cluttered and confusing GUI. i.e. if we press  [CTRL]-F the "Find" DB is shown and if next we press [CTRL]-H the new "Find&Replace" DB is displayed on top/next to the former Find DB, and so on. This is not good.

 b) Many times the action of calling the find dialog boxes is triggered by some text already marked in the current file; unfortunately Geany does not not always automatically copy the marked text as the text to look for ("Search for" field) in the invoked find dialog box.

After the patch
 a) Only one find dialog box can be displayed at the time. i.e If we want to "find text" we press [CTRL]-F, if later we want to "find and replace text" we press [CTRL]-H but now after this patch the "Find" dialog box will be closed before the "Find&Replace" DB is opened. This way the GUI is neater and far less distracting than before; it's only focused in the current find action.

 b) when opening any find dialog box the marked text in the file (if any) is "always automatically" copied to the "Search for" field of the invoked find dialog box avoiding unnecessary copy&paste actions.


Mod 2) 
Attaches to the tab menu a new option ("Copy Full Path") which copy to the clipboard the full path of the open file. 
This mod alters only the file /src/notebook.c. 
This mod is "very" handy when writing/commenting about code and we need a quick reference to the path of the particular file that we are probably describing a bug or a future mod, etc.


Please let me know if you guys need more detail about these mods.
Best,
Patrick

**Comment by codebrainz** [comment]
According to TravisCI this PR breaks the build.

**Comment by codebrainz** [comment]
> This mod is "very" handy when writing/commenting about code and we need a quick reference to the path of the particular file that we are probably describing a bug or a future mod, etc.

It sounds weird to be referring to an absolute path, would the comment be like this?

```c
/* This function is defined in C:\Users\John Smith\Projects\TheProject\Source\SomeModule.cc */
void foo(void);
```

**Comment by codebrainz** [comment]
I agree with @elextr this should be split into two PRs. For example I'd probably merge the notebook tab menu item one myself once it's fixed, but I probably won't touch it while it has all the unrelated stuff in the same PR.

It's pretty easy to split them:

```bash
$ git checkout master
$ git checkout -b notebook-copy-filename
$ git cherry-pick 656fbe657f176bda934cb5bc52f6715882ab7233
```

Something like that ought to make a new branch for the copy filename thing, then you can push that to your Github repo and make a new pull request.

**Comment by ppatpat** [comment]
Guys, I'm not using git on my builds, I just patch and rebuild the corresponding Debian package, the one I'm using "every" day; the mods work.
Later I paste the mods to github; then you can probably see "weird" formatting and things like that.

The mods are very simple but useful (at least for me) if you guys consider they have some value then feel free to alter the formatting, or the code, and test the thing. If you guys think they are not worth it just reject the PR.

Best,
Patrick

**Comment by elextr** [comment]
> Before the patch:
> a) The short-cuts
> [CTRL]-F, 
> [CTRL]-H, 
> [CTRL]-[SHIFT]-F 
> open the dialogs "Find", "Find & Replace", and "Find in files".
> All these dialog boxes can coexist "simultaneously" on top of the current file what produces a highly cluttered and confusing GUI. i.e. if we press [CTRL]-F the "Find" DB is shown and if next we press [CTRL]-H the new "Find&Replace" DB is displayed on top/next to the former Find DB, and so on. This is not good.

​IIUC this is intentional, they are modal dialogs, there is no reason for example that you should not be able to find-in-files, and then perform replacements alternately.  So I am against changing this.​  Better to be tidy and close the dialog when you are finished with it, click close or Alt+c.
 

> b) Many times the action of calling the find dialog boxes is triggered by some text already marked in the current file; unfortunately Geany does not not always automatically copy the marked text as the text to look for ("Search for" field) in the invoked find dialog box.

Works for me, if it doesn't for you please identify when and raise an issue/PR just for that fix.

[...]

> Mod 2) 
> Attaches to the tab menu a new option ("Copy Full Path") which copy to the clipboard the full path of the open file. 
> This mod alters only the file /src/notebook.c. 
> This mod is "very" handy when writing/commenting about code and we need a quick reference to the path of the particular file that we are probably describing a bug or a future mod, etc.

As @codebrainz says, this is a reasonable improvement, but it needs to be separated.

**Comment by elextr** [comment]
> Guys, I'm not using git on my builds, I just patch and rebuild the corresponding Debian package, the one I'm using "every" day; the mods work.

You need to at least submit the patch against the current development version, it basically does not compile due to an undefined variable, thats a pretty obvious problem, but who knows what subtle errors may occur by applying a patch from an old version.

**Comment by ppatpat** [comment]
>​IIUC this is intentional, they are modal dialogs, there is no reason for example that you should not be >able to find-in-files, and then perform replacements alternately. So I am against changing this.​ Better >to be tidy and close the dialog when you are finished with it, click close or Alt+c.
I find your way of working highly confusing, You can find-in-files then open among the found files the files you like and next perform a replacement **but never alternatively**. VS implements a highly mature IDE that has been tested by millions of developers during many (really many) years. I agree with them; find dialog boxes should be displayed one at the time. Just my 2 cents. BTW How many years have you been coding? 

>Works for me, if it doesn't for you please identify when and raise an issue/PR just for that fix.
No it doesn't work; please look at the code before you say things that are not true.
     /* only set selection if the dialog is not already visible */
     if (! gtk_widget_get_visible(replace_dlg.dialog) && sel)
the former code clearly shows that the selection is set **"only"** if the dialog is not visible; what IMHO is nuts.

>You need to at least submit the patch against the current development version, it basically does not >compile due to an undefined variable, thats a pretty obvious problem, but who knows what subtle >errors may occur by applying a patch from an old version.
1) guys, with all due respect, I do not "need" to at least submit anything. I just tried to share some code and I'm mostly getting comments on a missing space between "if" and "(", missing braces if the "if" is more than one line, and the like (pretty funny stuff BTW).
2) if there's a problem with a variable because of a version issue or probably the Debian package I'm using altering some thing, please let me know which is the undefined variable and I'll try to help with that. I took a look before posting the code and the structure was identical but probably I missed a changed variable name. At the end the mods are ultra simple it is a matter of 2 minutes for some one like you guys used to Geany code to see what I'm doing.

Best,
Patrick

**Comment by ppatpat** [comment]
I think the missing "page" variable issue is already solved. You guys have changed the prototype of
    src/notebook.c/static void show_tab_bar_popup_menu(GdkEventButton *event, GeanyDocument *doc)
and that lead to the error; Now it's fixed and works on v1.25

**Comment by codebrainz** [comment]
> guys, with all due respect, I just tried to share some code and I'm mostly getting comments on a missing space between "if" and "(", missing braces if the "if" is more than one line, and the like (pretty funny stuff BTW).

There's a coding convention we follow and if we just merged any old style that each contributor prefers the code would be even more of a mess. If you can't even bother to follow simple coding conventions, what are the chances that you paid attention to important stuff in your patches?

**Comment by ppatpat** [comment]
>There's a coding convention we follow and if we just merged any old style that each contributor prefers the code would be even more of a mess. 

I agree. But considering my patch is so simple, probably I expected talking more about the usefulness (or not) of the thing before than talking about a coding convention.

>If you can't even bother to follow simple coding conventions, what are the chances that you paid attention to important stuff in your patches?

You are mixing bananas with oranges; code quality is one thing and coding convention is a different matter. I have contributed code in many places in my life and I cannot consider strict coding conventions for a patch that has less than 100 lines. In other places they usually mention the coding convention when you become a frequent contributor, but never as the first comment when you just sent your first 90 lines of code..

Anyway, the patch works in 2.15 now, you can test it if you like. If you decide to promote the change and you really need some specific code convention trim please let me know.

Best,
Patrick

**Comment by codebrainz** [comment]
> You are mixing bananas with oranges; code quality/usefulness is one thing and coding convention is a different matter

It shows overall attention to detail and it's all described in detail in the HACKING file to avoid this kind of non-productive discussion.

> and I cannot consider strict coding conventions for a patch that has less than 100 lines

So because it's less than 100 lines, someone else should clean up the code for you? We're all volunteers, if you want to increase the likelihood of someone merging the change _you_ want, you should make it as easy as possible.

> but never as the first comment when you just sent your first 90 lines of code

I was doing a review of the changes, and it's the very first thing that stuck out to me and needs to be fixed before it can be merged, just as much as any missing symbols or other bugs.

**Comment by ppatpat** [comment]
>if you want to increase the likelihood of someone merging the change *you* want, you should make it as easy as possible

You are dead wrong; I **do not want** any change; I just thought of Geany's community; it is my way to contribute to a project I consider useful; BTW I get Geany **the way I want** in 5 minutes recompiling a Debian package w/o having to read comments on the number of spaces between an "if" and "("....

>So because it's less than 100 lines, someone else should clean up the code for you?

My friend.. this is not the kind of sentence that avoids escalating an argument... specially when  in my last post I wrote.

*If you decide to promote the change and you really need some specific code convention trim please let me know*

**Comment by elextr** [comment]
> the former code clearly shows that the selection is set only if the dialog is not visible; what IMHO is nuts.

Oh, you want the search string to continue to follow changes in selection.

No, to use your term, that is "nuts", you can't edit anything that involves changing the selection and keep the search term, which is especially bad when the user has set the search term to be a regular expression not just a simple string.

It is also "nuts" to apply arbitrary restrictions to what "search like" dialogs a user can have open, especially as find in files is completely unrelated to the current file search and replace dialogs.

And finally it is not acceptable to force changes that affect user workflows like this "because its the way VS works" without providing an option for them to turn the features off.

Copy filename to clipboard could be handy, and is likely to be acceptable as an individual PR.

**Comment by ppatpat** [comment]
>Oh, you want the search string to continue to follow changes in selection.

You are obviously not an every day coder using Geany. 
When working with any kind of "find" dialog box very often it's needed to perform "concatenated finds"; this means: 
1) We look for certain text_A 
2) When we find text_A we now see we need to find text_B and so on.
i.e. this is the classic case when parsing nested functions and we do not have intellisense available.
In that situation after finding text_A it should only take "selecting" the text_B on the doc and just pressing [CTRL]-F again for getting the new marked text (text_B) automatically loaded in the find dialog box. This would avoid the unnecessary copy&paste actions improving productivity. Can you see it now? Note: VS, Eclipse, Netbeans, etc all of them implement this feature. Geany doesn't.

>No, to use your term, that is "nuts", you can't edit anything that involves changing the selection and keep the search term, which is especially bad when the user has set the search term to be a regular expression not just a simple string.

What???

>It is also "nuts" to apply arbitrary restrictions to what "search like" dialogs a user can have open, especially as find in files is completely unrelated to the current file search and replace dialogs.

Again you prove here that you are talking about what you do not use regularly. 
1. Do you think "Find" and Find&replace" make sense when displayed at the same time? off course not. Note: Eclipse even has those dialogs as only "one", Netbeans displays only one at the time as a search bar, VS switches between one or the other one. Geany displays both at the same time...

2. I have already explained you why "Find" (or Find&replace" ) also makes no sense when displayed with "Find in Files"; if interested read the corresponding posts above. Note: Eclipse does not allow to use both styles of dialog boxes at the same time.

3. Finally, please let all of us know if you really code every day using Geany (or any serious IDE for that matter)

>And finally it is not acceptable to force changes that affect user workflows like this "because its the way VS works" without providing an option for them to turn the features off.

Please.. who's the one that is forcing anything?? 
The find dialog box style proposed here is the neater approach I've ever found; "coincidentally" is the one implemented by VS, Eclipse, etc. In case you wonder yes, I'm an every day coder using Geany, VS, etc, etc. 

I would agree though that a configuration parameter switching the dialog boxes behavior could be a good idea for people like you used to cumbersome ways to move around an IDE. If you like that idea then you have homework.

Best,
Patrick

**Comment by techee** [comment]
@ppatpat I think you took a very unfortunate confrontational approach which will eventually lead to not merging your patches not because they aren't useful but because others will get tired of the pointless discussion like above. Here are a few points related to my experience with how to get my patches merged:

**Do a good patch marketing.** It's your task to convince others your patch is useful - the fact that the particular feature isn't in Geany yet probably means others don't care much and you must convince them it's something worth having in Geany. Clearly describe what problem you are trying to solve with your patches - what the state was before and how your patch improves things. Do this preferably directly in the commit message.

Until your fifth comment in this thread it wasn't clear to me what you are trying to achieve apart from "doing it the VS way" (I have never used visual studio and the fact that another editor does things differently isn't a good argument). By the way Gedit works the way you suggest too which is a closer Geany's competitor.

**Save the reviewer's/maintainer's time.** It's the maintainer's time which is expensive, not yours. There are typically tens of patches waiting for review and the maintainer will pick those which are as easy to review as possible. You significantly increase the likelihood your patches get reviewed if they are nice-looking - don't submit a half-baked crap for review.

IMO your patches are really messy. First, they don't contain any commit messages. Describe what your are trying to achieve as suggested above and the way you are implementing it. Patches like "Update notebook.c" or "Update search.c" really don't help the reviewer to identify what you perform in the patch. Patches should describe the particular feature you are trying to implement, not files which got changed. What might help is looking at already merged patches to see what form they should have.

Second, do a self-review of the code you publish - don't leave commented-out code in the patch, try to follow the style you see in the code as much as possible (even if you feel your style is the "right" one)

**Take the "the reviewer is always right" approach** (unless it's something important you disagree with). You can start arguing your coding style is better, that the review process should change, that other open source projects do something in a different ("better") way etc. but these things just move the discussion somewhere else and delay the patch acceptance. Concentrate on the important thing - to get the patch in and don't lose your focus. Typically it's also faster to update the patches than endless discussions.

For instance, does the reviewer say you should drop curly braces around single-line if's? Simply do it. You can start a discussion that you have been using curly braces in this case because it's safer, that it's considered a best-practice and that other projects do it this way too but you just start fragmenting the discussion and instead of getting your patch in you get stuck on something which is irrelevant to the patch and the patch doesn't get merged. (If you want to change the coding style, open a new ticket or suggest something on the mailing list).

And really, really don't start getting unfriendly towards others: accusing long-term Geany developers of "not being an every day coder using Geany", telling people they are "mixing bananas with oranges" when they just say your code should be cleaned up a bit before it can be reviewed and so on is plain stupid. This kind of "I know best, others are idiots" approach has never helped anything and nobody wants to work with people with such an attitude.

**Be determined to finish what you have started.** I must say I was quite fascinated by your reaction here:

>>if you want to increase the likelihood of someone merging the change you want, you should make it as easy as possible

>You are dead wrong; I do not "want" any change; I just thought of Geany's community; it is my way to contribute to a project I consider useful; BTW I get Geany the way I want in 5 minutes recompiling a Debian package w/o having to read "funny" comments on the number of spaces between an "if" and "("....

What do you want then? In my opinion submitting a patch for review means "I want to get this change into Geany and I will continue working on the patches based on the reviewer's comments". If you don't want to get the patch into Geany, it's pointless to submit it.

And be prepared that during the review process there will be further comments requiring additional work - for my patches it's often something like 40% time spent on the initial set of patches and 60% to finally get them in. If you are not willing to incorporate the reviewer's comments, nobody else will do the work for you.

**Finally, be patient.** I sometimes have problems with this myself but there's a limited time people involved in Geany can spend so your patches can be unreviewed for quite some time.

---

tl;dr
What the patch tries to achieve seems reasonable to me; the overall attitude on the other hand looks like from a book with a title "the best approaches not to get my patches merged".

By the way, I have never seen the problem the patch tries to solve myself because I use the "Hide the find dialog" option from Preferences->General->Miscellaneous so it's always hidden after searching so there are never 2 diallogs at the same time and the marked text is always preloaded because Ctrl+F leads to opening the dialog.

---

One more note: I don't want to turn this into some kind of flame war so I'm probably not going to continue in the discussion if it looks similarly to what preceded this post.

**Comment by ppatpat** [comment]
@techee I appreciate you taking the time. 

It seems to me your motivations for publishing your code are different than mine. As I've said before I use Geany, I improved a couple of things and I just wanted to make them public. That's my way to say "thank you" to the community that makes Geany. Making my changes available to the public is all I need here; I do not really need to see them "accepted" nor "implemented" either.

>It's the maintainer's time which is expensive, not yours... IMO your patches are really messy.... 

Sorry; I do not agree.

>Take the "the reviewer is always right" approach

I have been reviewing code myself many times and this is my top-down approach.
1. First I **must** say **Thank you** on my first contact (even if the patch is pure crap).
2. Next I discuss **Functionality**; what does the patch do? does it overlap/integrate correctly? is it any good?
3. Next I discus the **Implementation**; Does the patch use the right APIs? the right algorithms? core vs plugin? that sort of things...
4. Finally I discuss the **Formal** aspects; Does the patch follow the coding convention? should the patch be split? does the patch include the right license terms? etc

As reviewer I try to be **assertive**; 
1. A reviewer that says i.e. *I don't think user_data needs a cast here.* is telling me he/she has **no idea** but he/she **suspects the poster is wrong**.  If I'm reviewing code and I find myself in doubt I must find out what's going on before opening my mouth or I just remain silent.
2. A reviewer that talks about things he/she has probably no much experience with should be very careful before trying to be assertive... 

As you can see in this case we started discussing the **formal** aspects of the thing while skipping the **fundamental** point 1.

>By the way, I have never seen the problem the patch tries to solve myself because I use the "Hide the find dialog" option from Preferences->General->Miscellaneous so it's always hidden after searching so there are never 2 diallogs at the same time 

That option **closes** the find dialog box after a match. That is not really a very handy feature. The find dialog should remain open while not purposely closed (that's what most serious IDEs do).

>and the marked text is always preloaded because Ctrl+F leads to opening the dialog

This **only** happens when the find dialog is not previously displayed. Please read my description of "concatenated searches".

>I don't want to turn this into some kind of flame war

Sure; me neither.

Best,
Patrick

**Comment by codebrainz** [comment]
On 15-07-30 01:02 AM, ppatpat wrote:
> @techee I appreciate you taking the time.
>
> It seems to me your motivations for publishing your code are different than mine. As I've said before I use Geany, I improved a couple of things and I just wanted to make them public. That's my way to say "thank you" to the community that makes Geany. Making my changes available to the public is all I need here; I do not really need to see them "accepted" nor "implemented" either.
>

If you haven't bothered to make them acceptable, and you don't care if 
they're accepted, why would you make a Pull Request? If you just want to 
share random stuff, leave it on your fork of the repo. A pull request 
tells the maintainers that you're interested in having this reviewed and 
hopefully merged into core and that you've made an effort to make it 
ready for that.

If you just want to toss around ideas, use the mailing list, or if you 
think something is a bug, open a bug report.

>> It's the maintainer's time which is expensive, not yours... IMO your patches are really messy....
>
> Sorry; I do not agree.
>

Everyone's time is important, and your patches are really messy. None of 
us get paid to contribute to open source, so if you want some feature 
in, the best thing to do is make the patches clean and reviewable, 
building against the branch to merged into, following the coding 
conventions, and not being a jerk to the people volunteering their time 
to help _you_ get your change in.

>> Take the "the reviewer is always right" approach
>
> I have been reviewing code myself many times and this is my top-down approach.
> 1. First I **must** say **Thank you** on my first contact (even if the patch is pure crap).

**Thank you** for wasting my time on pure crap patches that I don't 
personally give a darn about and was only doing you a favour to try and 
review and merge something you didn't even bother to make nice or care 
if it's merged. Also the abrasive, insulting and off-topic discussion 
has been a treat.

> 2. Next I discuss **Functionality**; what does the patch do? does it overlap/integrate correctly? is it any good?
> 3. Next I discus the **Implementation**; Does the patch use the right APIs? the right algorithms? core vs plugin? that sort of things...

It was so simple and non-controversial a patch, that even though it was 
mixed in with unrelated stuff in the PR, I was going to just quickly 
check it and cherry-pick it, had it been in any kind of shape to do so, 
without any further discussion.

> 4. Finally I discuss the **Formal** aspects; Does the patch follow the coding convention? should the patch be split? does the patch include the right license terms? etc
>

If you just do this before opening a pull request, there's no silly 
discussion about spaces around parenthesis or future noisy commits 
fixing trivial stuff like white space. Even a complete new-comer can 
figure this part out and make it a non-issue.

> As reviewer I try to be **assertive**;
> 1. A reviewer that says i.e. *I don't think user_data needs a cast here.* is telling me he/she has **no idea** but he/she **suspects the poster is wrong**.  If I'm reviewing code and I find myself in doubt I must find out what's going on before opening my mouth or I just remain silent.

Sorry, we have pull requests from people with VASTLY different skill 
levels, from those who wrote their first line of C code in that very 
pull request all the way to those who have been programming in C since 
before I was born.

What I was trying to say, translated from Canadian to American is "Why 
the hell are you doing a dynamic cast from a void-pointer to a 
GtkWidget, this isn't C++, what are you thinking!?" except the way I 
worded it won't make you feel like an idiot if you're a n00b and it 
leaves the possibility for you to rebut with a real explanation if you 
you know better.

> 2. A reviewer that talks about things he/she has probably no much experience with should be very careful before trying to be assertive...
>

Why do you talk like this? Please speak plainly. If you want to call me 
inexperienced why don't you just say that? I never claimed to be a 
know-it-all, I'm just a guy who likes programming and contributed enough 
good stuff to prove my worth on the team. I'm not going be condescending 
to everyone and loudly proclaiming I know more than them because I have 
commit rights.

> As you can see in this case we started discussing the **formal** aspects of the thing while skipping the **fundamental** point 1.
>

Again, the patch I reviewed, even though it was muddled in an unrelated 
PR was so trivial that it didn't require any of that. All it needed was 
to follow the coding convention and make sure it didn't break the build.

>> [...]
>> I don't want to turn this into some kind of flame war
>
> Sure; me neither.
>

To that end, I'm going to close this pull request. If you actually want 
the changes merged, please open a separate PR for each topic, coded to 
the best of your abilities, and don't come out with guns blazing. The 
Geany community gets along fine being civil to each other, at least to date.

**Comment by ppatpat** [comment]
>If you haven't bothered to make them acceptable, 

This is not true; there was an obvious issue with v1.25 the one that was immediately solved. I have also said that if the patch was considered for inclusion I was willing to parse the coding convention issues; then please stick to the facts.

>and you don't care if they're accepted, why would you make a Pull Request?

Please do not change my words; **I do care** about the inclusion of my code but that is not my main goal here. Can you see the difference? 
I've made a pull request just to expose my code; leaving it in my repo is like keeping it on my HDD; The Pull Request makes people like **you** to know about my code and surely some day someone will add these or similar mods; Trust me it's just a matter of time.

>If you just want to toss around ideas, use the mailing list, or if you think something is a bug, open a bug report.

Tossing around ideas is good when you haven't coded a single line; I have a fully working mod; can you see the difference?

>Everyone's time is important, and your patches are really messy. None of us get paid to contribute to open source, so if you want some feature in, the best thing to do is make the patches clean and reviewable, building against the branch to merged into, following the coding conventions, and not being a jerk to the people volunteering their time to help _you_ get your change in.

I appreciate if you do not use personal adjectives; keep it professional please.

>**Thank you** for wasting my time on pure crap patches that I don't personally give a darn about and was only doing you a favour to try and review and merge something you didn't even bother to make nice or care if it's merged. Also the abrasive, insulting and off-topic discussion has been a treat.

Enough my friend; you are loosing your temper; that's not good. This conversation ends here.

Best,
Patrick

**Comment by FProchaska** [comment]
@codebrainz, @elextr, sad to see how you refuse to help users and community to contribute to your code base they basically do not know anything about. I read this as "We wish this tool to stand out the way Vim does. Who wants modern editor please go use VS Code or Notepad++."  Very unwelcoming towards a new contributor and also to users. BTW, for keeping search phrase the editors usually have F3 as a shortcut. So sad to see this.
